### PR TITLE
Allow a custom API key

### LINKED ISSUE #36: How to pass API key
I want to add API key . Where I can add it ?

### LINKED ISSUE #37: Basic test failed
Can someone check if this module is being maintained.
Basic test failed with latest version and works fine with 6.x

```
const ipLocation = require("iplocation");

(async () => {
    await ipLocation("172.217.167.78");
})();
```

With 7.x (latest), i see the following error
HTTPError: Too Many Requests

### PR DESCRIPTION
**Pull request by Aman1997**
Fixes #36, Fixes #37

### CODE REVIEW

File: source/index.ts [hunk 1]
```diff
@@ -83,52 +83,82 @@ const ipLocation = require("ip-location");
 })();
 ```
 */
-async function ipLocation(ip: string): Promise<ipLocation.ReturnType> {
+async function ipLocation(
+	ip: string,
+	api_key: string = ""
```
Comment (Richienb) [inline source/index.ts]: Camel case should be used

Comment (Richienb) [inline source/index.ts]: Should be an optional argument

File: source/index.ts [hunk 2]
```diff
@@ -83,52 +83,82 @@ const ipLocation = require("ip-location");
...
+		in_eu,
+		postal,
+		timezone,
+		utc_offset,
+		currency,
+		currency_name,
+		languages,
+		country_area,
+	}: IpApiData = await ky(
+		`https://ipapi.co/${ip}/json/${api_key ? `?key=${api_key}` : ""}`
```
Comment (Richienb) [inline source/index.ts]: You can use `searchParams` instead of encoding it directly

**Review by Richienb — CHANGES_REQUESTED** [review]
Some auto-formatting needs to be reverted

File: source/index.ts (hunk 1 above)
Comment (Aman1997) [reply source/index.ts]: Updated

Comment (Aman1997) [reply source/index.ts]: Updated

File: source/index.ts (hunk 2 above)
Comment (Aman1997) [reply source/index.ts]: Conditionally adding the api key to the URL as searchParams.

File: source/index.ts [hunk 3]
```diff
@@ -83,23 +83,29 @@ const ipLocation = require("ip-location");
 })();
 ```
 */
-async function ipLocation(ip: string): Promise<ipLocation.ReturnType> {
+async function ipLocation(ip: string, apiKey?: string): Promise<ipLocation.ReturnType> {
```
Comment (Richienb) [inline source/index.ts]: This should still be in an options object

File: source/index.ts [hunk 4]
```diff
@@ -54,7 +54,7 @@ declare namespace ipLocation {
 			currency: {
 				name: string
 				code: string
-			}
+			};
```
Comment (Richienb) [inline source/index.ts]:
```suggestion
			}
```

File: source/index.ts [hunk 5]
```diff
@@ -83,23 +83,29 @@ const ipLocation = require("ip-location");
...
 	if (typeof ip !== "string" || !isIp.v4(ip)) {
 		throw new TypeError("A valid ipv4 address must be provided!")
 	}
 
-	const { latitude, longitude, city, reserved, region, region_code, country_name, country_code, country_code_iso3, country_capital, country_tld, country_population, country_calling_code, continent_code, in_eu, postal, timezone, utc_offset, currency, currency_name, languages, country_area }: IpApiData = await ky(`https://ipapi.co/${ip}/json/`).json()
+	const IPAPI_URL = new URL(`https://ipapi.co/${ip}/json/`)
+
+	if (apiKey) {
+		IPAPI_URL.searchParams.append("key", apiKey)
+	}
```
Comment (Richienb) [inline source/index.ts]: By searchParams option, I mean you can pass it as the second argument to [ky](https://github.com/sindresorhus/ky)

File: source/index.ts [hunk 6]
```diff
@@ -83,23 +83,29 @@ const ipLocation = require("ip-location");
...
+
+	const { latitude, longitude, city, reserved, region, region_code, country_name, country_code, country_code_iso3, country_capital, country_tld, country_population, country_calling_code, continent_code, in_eu, postal, timezone, utc_offset, currency, currency_name, languages, country_area }: IpApiData = await ky(IPAPI_URL.href).json()
 
 	return reserved ? {
-		reserved
+		reserved,
```
Comment (Richienb) [inline source/index.ts]:
```suggestion
		reserved
```

File: source/index.ts [hunk 7]
```diff
@@ -83,23 +83,29 @@ const ipLocation = require("ip-location");
...
 	} : {
 		latitude,
 		longitude,
 		city,
 		reserved: Boolean(reserved),
 		region: {
 			name: region,
-			code: region_code
+			code: region_code,
```
Comment (Richienb) [inline source/index.ts]:
```suggestion
			code: region_code
```

File: source/index.ts [hunk 8]
```diff
@@ -119,7 +125,7 @@ async function ipLocation(ip: string): Promise<ipLocation.ReturnType> {
 				name: currency_name,
 				code: currency
 			},
-			languages: languages ? languages.split(",") : []
+			languages: languages ? languages.split(",") : [],
```
Comment (Richienb) [inline source/index.ts]:
```suggestion
			languages: languages ? languages.split(",") : []
```

**Review by Richienb — CHANGES_REQUESTED** [review]

File: source/index.ts (hunk 3 above)
Comment (Aman1997) [reply source/index.ts]: Made apiKey parameter as part of an optional options object

File: source/index.ts (hunk 5 above)
Comment (Aman1997) [reply source/index.ts]: Updated

File: source/index.ts [hunk 9]
```diff
@@ -83,12 +87,12 @@ const ipLocation = require("ip-location");
...
 ```
 */
-async function ipLocation(ip: string): Promise<ipLocation.ReturnType> {
+async function ipLocation(ip: string, options?: ipLocation.Options): Promise<ipLocation.ReturnType> {
 	if (typeof ip !== "string" || !isIp.v4(ip)) {
 		throw new TypeError("A valid ipv4 address must be provided!")
 	}
 
-	const { latitude, longitude, city, reserved, region, region_code, country_name, country_code, country_code_iso3, country_capital, country_tld, country_population, country_calling_code, continent_code, in_eu, postal, timezone, utc_offset, currency, currency_name, languages, country_area }: IpApiData = await ky(`https://ipapi.co/${ip}/json/`).json()
+	const { latitude, longitude, city, reserved, region, region_code, country_name, country_code, country_code_iso3, country_capital, country_tld, country_population, country_calling_code, continent_code, in_eu, postal, timezone, utc_offset, currency, currency_name, languages, country_area }: IpApiData = await ky(`https://ipapi.co/${ip}/json/`, { searchParams: options?.apiKey ? { 'key': options.apiKey } : null }).json()
```
Comment (Richienb) [inline source/index.ts]: You can just reference apiKey directly and if it resolves to undefined, ky will automatically ignore it

**Review by Richienb — CHANGES_REQUESTED** [review]
The readme should be updated

File: source/index.ts (hunk 9 above)
Comment (Aman1997) [reply source/index.ts]: okay

File: source/index.ts [hunk 10]
```diff
@@ -83,12 +87,12 @@ const ipLocation = require("ip-location");
...
 ```
 */
-async function ipLocation(ip: string): Promise<ipLocation.ReturnType> {
+async function ipLocation(ip: string, options?: ipLocation.Options): Promise<ipLocation.ReturnType> {
 	if (typeof ip !== "string" || !isIp.v4(ip)) {
 		throw new TypeError("A valid ipv4 address must be provided!")
 	}
 
-	const { latitude, longitude, city, reserved, region, region_code, country_name, country_code, country_code_iso3, country_capital, country_tld, country_population, country_calling_code, continent_code, in_eu, postal, timezone, utc_offset, currency, currency_name, languages, country_area }: IpApiData = await ky(`https://ipapi.co/${ip}/json/`).json()
+	const { latitude, longitude, city, reserved, region, region_code, country_name, country_code, country_code_iso3, country_capital, country_tld, country_population, country_calling_code, continent_code, in_eu, postal, timezone, utc_offset, currency, currency_name, languages, country_area }: IpApiData = await ky(`https://ipapi.co/${ip}/json/`, { searchParams: { 'key': options?.apiKey } }).json()
```
Comment (Richienb) [inline source/index.ts]: The quotes around `key` are optional.

File: README.md [hunk 11]
```diff
@@ -15,22 +15,39 @@ npm install iplocation
...
 (async () => {
 	await ipLocation("172.217.167.78");
 	//=> { latitude: -33.8591, longitude: 151.2002, region: { name: "New South Wales" ... } ... }
 })();
+
+// using the paid tier of ipapi
+(async () => {
+	await ipLocation("172.217.167.78", { apiKey: 'YOUR_API_KEY'});
+	//=> { latitude: -33.8591, longitude: 151.2002, region: { name: "New South Wales" ... } ... }
+})();
```
Comment (Richienb) [inline README.md]: This example can be moved directly underneath the description of the `apiKey` option.

File: README.md [hunk 12]
```diff
@@ -15,22 +15,39 @@ npm install iplocation
...
 
+#### options
+
+Type: `object`
+
+#### apiKey
+
+Type: `string`
+
+The api key for the paid version of [ipapi](https://ipapi.co/pricing/?utm_source=google&utm_term=ipapi&utm_campaign=10951547418&adgroupid=109238001284&device=c&utm_medium=cpc&utm_content=459469000491&gclid=Cj0KCQjwiNSLBhCPARIsAKNS4_dAe4tHfaLyWSQSqp0kZUnZBrhTkldv5cRYLMqH_XinJuyuJ8Dplg0aAmrsEALw_wcB).
```
Comment (Richienb) [inline README.md]:
```suggestion
An optional [ipapi](https://ipapi.co/) API key to use.
```

File: source/index.ts [hunk 13]
```diff
@@ -67,6 +67,10 @@ declare namespace ipLocation {
 		reserved: boolean
 	}
 
+	export interface Options { 
+		apiKey: string
```
Comment (Richienb) [inline source/index.ts]: The description and example should be added here as well in a multi-line comment.

**Review by Richienb — CHANGES_REQUESTED** [review]

File: source/index.ts (hunk 13 above)
Comment (Aman1997) [reply source/index.ts]: Added the new example within the existing commented sample code

Comment (Richienb) [reply source/index.ts]: Not only should the example be in the readme, but it should also be here as well. [Example](https://github.com/sindresorhus/ky/blob/2049affe2ca939f53350fb9b7354c3ff8d2e1c1f/source/types/options.ts#L38-L73).

File: README.md [hunk 14]
```diff
@@ -23,14 +24,32 @@ const ipLocation = require("iplocation");
...
+Type: `object`
+
+#### apiKey
+
+Type: `string`
+
+An optional [ipapi](https://ipapi.co/) API key to use.
+
+```js
+// using the paid tier of ipapi
```
Comment (Richienb) [inline README.md]: These comments should be removed and the import statement should be added here.

File: README.md [hunk 15]
```diff
@@ -23,14 +23,33 @@ const ipLocation = require("iplocation");
...
+
+Type: `string`
+
+An optional [ipapi](https://ipapi.co/) API key to use.
+
+```js
+const ipLocation = require("iplocation");
+
+(async () => {
+	await ipLocation("172.217.167.78", { apiKey: 'YOUR_API_KEY'});
```
Comment (Richienb) [inline README.md]:
```suggestion
	await ipLocation("172.217.167.78", { apiKey: 'YOUR_API_KEY' });
```

File: source/index.ts (hunk 13 above)
Comment (Aman1997) [reply source/index.ts]: The example of how to use apiKey is added to the main index.ts file below the existing example

Comment (Richienb) [reply source/index.ts]: That's not where it goes in this codebase. It should be put above the option.

Comment (Aman1997) [reply source/index.ts]: okay!

Comment (Richienb) [reply source/index.ts]: Check the example again. You need to create a comment directly above `apiKey` and add the description and example of the option specifically in there which you can grab from the readme.

**Comment by Richienb** [comment]
Thanks!

**Comment by vishwasnavadak** [comment]
@Richienb Will there be a NPM release with this fix anytime soon?

**Pull request merged**